scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala

```diff
@@ -77,5 +64,5 @@ private[values] class MultiMapSideInput[K, V](val view: PCollectionView[JMap[K,
 /** Encapsulate context of one or more [[SideInput]]s in an [[SCollectionWithSideInput]]. */
 class SideInputContext[T] private[scio] (val context: DoFn[T, AnyRef]#ProcessContext) {
   /** Extract the value of a given [[SideInput]]. */
-  def apply[S](side: SideInput[S]): S = side.getCache(context)
+  def apply[S](side: SideInput[S]): S = side.get(context)
```

Comment:
Would this be potentially expensive for `List`, `Iterable` and `Map`s?

Reply:
We call `asScala` for list, iterable, map, and multi-map but all of them should be just a thin wrapper. There's some overhead in creating the wrapper objects but hopefully JVM can optimize that.

Reply:
Is this sth we could verify/measure, cause this would potentially be called for each element, right?

Reply:
Gonna run some benchmark to find out.

Reply:
Without cache:
```
================================================================================
Benchmark           : SingletonSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:44:50.187017Z
Finish time         : 2017-10-06T16:49:17.033686Z
Elapsed             : 266 seconds
TotalMemoryUsage    : 10445106
TotalPdUsage        : 170004
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 2720
================================================================================
Benchmark           : IterableSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:44:53.243249Z
Finish time         : 2017-10-06T16:48:42.952200Z
Elapsed             : 229 seconds
TotalMemoryUsage    : 8601669
TotalPdUsage        : 140001
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 2240
================================================================================
Benchmark           : ListSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:44:56.550629Z
Finish time         : 2017-10-06T16:48:37.497596Z
Elapsed             : 220 seconds
TotalMemoryUsage    : 7987252
TotalPdUsage        : 130000
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 2080
================================================================================
Benchmark           : MapSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:44:59.631426Z
Finish time         : 2017-10-06T16:47:52.942432Z
Elapsed             : 173 seconds
TotalMemoryUsage    : 4300824
TotalPdUsage        : 70000
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 1120
================================================================================
Benchmark           : MultiMapSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:45:02.556561Z
Finish time         : 2017-10-06T16:47:46.134817Z
Elapsed             : 163 seconds
TotalMemoryUsage    : 4300827
TotalPdUsage        : 70000
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 1120
```

Reply:
With cache:
```
================================================================================
Benchmark           : SingletonSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:50:08.823460Z
Finish time         : 2017-10-06T16:53:52.973031Z
Elapsed             : 224 seconds
TotalMemoryUsage    : 8601732
TotalPdUsage        : 140002
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 2240
================================================================================
Benchmark           : IterableSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:50:11.638010Z
Finish time         : 2017-10-06T16:53:26.844235Z
Elapsed             : 195 seconds
TotalMemoryUsage    : 6758439
TotalPdUsage        : 110000
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 1760
================================================================================
Benchmark           : ListSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:50:14.201950Z
Finish time         : 2017-10-06T16:53:36.639266Z
Elapsed             : 202 seconds
TotalMemoryUsage    : 7372852
TotalPdUsage        : 120000
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 1920
================================================================================
Benchmark           : MapSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:50:16.921297Z
Finish time         : 2017-10-06T16:53:16.662730Z
Elapsed             : 179 seconds
TotalMemoryUsage    : 4915234
TotalPdUsage        : 80000
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 1280
================================================================================
Benchmark           : MultiMapSideInput
Extra arguments     : 
State               : DONE
Create time         : 2017-10-06T16:50:19.709772Z
Finish time         : 2017-10-06T16:53:08.705675Z
Elapsed             : 168 seconds
TotalMemoryUsage    : 4915241
TotalPdUsage        : 80000
TotalShuffleUsage   : 0
TotalSsdUsage       : 0
TotalVcpuTime       : 1280
```

Reply:
@nevillelyh what are the sizes of main stream and side inputs for these tests?

Reply:
main 100M, side 1M for singleton/iterable/list
main 1M, side 100K for map